tempesta_fw/http_msg.c

```diff
@@ -1043,3 +1047,146 @@ __tfw_http_msg_alloc(int type, bool full)
 	return hm;
 }
 
+/**
+ * Determination length of the header's real part (for details see comment
+ * for @tfw_http_msg_hdr_write() below) to store it in the encoder dynamic
+ * index.
+ */
+unsigned long
+tfw_http_msg_hdr_length(const TfwStr *hdr, unsigned long *name_len,
+			unsigned long *val_off, unsigned long *val_len)
+{
+	const TfwStr *chunk, *end;
+	unsigned long tail, hdr_tail = 0, hdr_len = 0;
+	bool name_found = false, val_found = false;
+
+	TFW_STR_FOR_EACH_CHUNK(chunk, hdr, end) {
+		unsigned long idx;
+
+		if (!chunk->len)
+			continue;
+
+		idx = chunk->len - 1;
+		hdr_len += chunk->len;
+		if (!name_found) {
+			for (; (chunk->data[idx] == ' '
+				|| chunk->data[idx] == '\t')
+				     && idx;
+			     --idx);
+
+			*name_len += idx;
```

Comment:
Yes, agree - too tricky place. Due to https://github.com/tempesta-tech/tempesta/pull/1338#discussion_r320937268 and corresponding discussion in chat, this function will be reworked in order to avoid complexity (and through the consensual planned refinements of the current HTTP parser).

Reply:
ack